java/src/main/java/com/google/oak/transport/GrpcStreamingTransport.java

```diff
@@ -0,0 +1,150 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package com.google.oak.transport;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import com.google.oak.transport.EvidenceProvider;
+import com.google.oak.transport.Transport;
+import com.google.oak.util.Result;
+import com.google.protobuf.ByteString;
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.stub.StreamObserver;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.function.Function;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import oak.session.noninteractive.v1.AttestationBundle;
+import oak.session.noninteractive.v1.GetPublicKeyRequest;
+import oak.session.noninteractive.v1.GetPublicKeyResponse;
+import oak.session.noninteractive.v1.InvokeRequest;
+import oak.session.noninteractive.v1.InvokeResponse;
+import oak.session.noninteractive.v1.RequestWrapper;
+import oak.session.noninteractive.v1.ResponseWrapper;
+
+public class GrpcStreamingTransport implements EvidenceProvider, Transport {
+  private static final Logger logger = Logger.getLogger(GrpcStreamingTransport.class.getName());
+
+  private static final int MESSAGE_QUEUE_TIMEOUT_SECONDS = 10;
+
+  private class ResponseStreamObserver implements StreamObserver<ResponseWrapper> {
+    @Override
+    public void onNext(ResponseWrapper response) {
+      try {
+        messageQueue.put(response);
+      } catch (Exception e) {
+        if (e instanceof InterruptedException) {
+          Thread.currentThread().interrupt();
+        }
+        logger.log(Level.WARNING, "Couldn't send server response to the message queue: {0}", e);
+      }
+    }
+
+    @Override
+    public void onError(Throwable t) {
+      Status status = Status.fromThrowable(t);
+      logger.log(Level.WARNING, "Couldn't receive response: {0}", status);
+    }
+
+    @Override
+    public void onCompleted() {
+      logger.log(Level.FINE, "response message queue completed");
+    }
+  }
+
+  private BlockingQueue<ResponseWrapper> messageQueue;
+  private StreamObserver<RequestWrapper> requestObserver;
+
+  /**
+   * Creates an instance of {@code GrpcStreamingTransport}.
+   *
+   * @param stream a method reference to a gRPC client streaming method with the appropriate request
+   * and response types.
+   */
+  public GrpcStreamingTransport(
+      Function<StreamObserver<ResponseWrapper>, StreamObserver<RequestWrapper>> stream) {
+    // We expect to receive a single response message for each request.
+    this.messageQueue = new ArrayBlockingQueue<>(1);
```

Comment:
For better readability, you can do the initialization where the variable is declared, since it does not use anything from the constructor inputs.

Reply:
Done